File: app/src/main/java/com/vocaby/app/api/ApiManager.java
package com.vocaby.app.api;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.vocaby.app.Constants;
import com.vocaby.app.models.datapackage.EntryDataPackage;

import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiManager {
    final String BASE_URL = Constants.VOCABY_API_SERVER;
    private static ApiManager instance;
    private static VocabyApiService apiService;

    private ApiManager() {
    }

    public static synchronized ApiManager getInstance() {
        if (null == instance) {
            instance = new ApiManager();
        }

        return instance;
    }

    public VocabyApiService getVocabyApiService(int type) {
        apiService = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(createGsonConverterFactory(type))
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .build()
                .create(VocabyApiService.class);

        return apiService;
    }

    private static GsonConverterFactory createGsonConverterFactory(int type) {
        GsonBuilder gsonBuilder = new GsonBuilder();
        // Adding custom deserializers
        switch(type) {
            case VocabyApiService.DEFINITION:
                gsonBuilder.registerTypeAdapter(EntryDataPackage.class, new GetWordDataDeserializer());
                break;
            case VocabyApiService.LOGIN:
                gsonBuilder.registerTypeAdapter(AuthResponse.class, new GetAuthDeserializer());
                break;
            default:
                return GsonConverterFactory.create();
        }

        Gson myGson = gsonBuilder.create();

        return GsonConverterFactory.create(myGson);
    }
}

File: app/src/main/java/com/vocaby/app/api/ApiManager.kt
package com.vocaby.app.api

import com.google.gson.GsonBuilder
import com.vocaby.app.Constants.VOCABY_API_BASE_URL
import com.vocaby.app.models.dictionary.EntryModel
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit


object ApiManager {
    val apiService: ApiService by lazy {
        val gsonBuilder = GsonBuilder()
        gsonBuilder.registerTypeAdapter(EntryModel::class.java, EntryDeserializer())
        val vocabyGson = gsonBuilder.create()
//        val httpLoggingInterceptor = HttpLoggingInterceptor()
//        httpLoggingInterceptor.apply {
//            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
//        }

        val okHttpClient = OkHttpClient.Builder()
//            .addInterceptor(httpLoggingInterceptor)
            .connectTimeout(3, TimeUnit.SECONDS)
            .readTimeout(3, TimeUnit.SECONDS)
            .writeTimeout(3, TimeUnit.SECONDS)
            .build()

        Retrofit.Builder()
            .baseUrl(VOCABY_API_BASE_URL)
            .addConverterFactory(GsonConverterFactory.create(vocabyGson))
            .client(okHttpClient)
            .build()
            .create(ApiService::class.java)
    }
}